xds/internal/balancer/ringhash/e2e/ringhash_balancer_test.go

```diff
@@ -949,18 +929,691 @@ func (s) TestRingHash_EndpointWeights(t *testing.T) {
+		ServiceName: clusterName,
+		Policy:      e2e.LoadBalancingPolicyRingHash,
+	})
+	route := headerHashRoute("new_route", virtualHostName, clusterName, "address_hash")
+	listener := e2e.DefaultClientListener(virtualHostName, route.Name)
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	xdsServer, nodeID, xdsResolver := setupManagementServerAndResolver(t)
+	if err := xdsServer.Update(ctx, xdsUpdateOpts(nodeID, endpoints, cluster, route, listener)); err != nil {
+		t.Fatalf("Failed to update xDS resources: %v", err)
+	}
+
+	conn, err := grpc.NewClient("xds:///test.server", grpc.WithResolvers(xdsResolver), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("Failed to create client: %s", err)
+	}
+	defer conn.Close()
+	client := testgrpc.NewTestServiceClient(conn)
+
+	// Note each type of RPC contains a header value that will always be hashed
+	// the value that was used to place the non-existent endpoint on the ring,
+	// but it still gets routed to the backend that is up.
+	ctx = metadata.NewOutgoingContext(ctx, metadata.Pairs("address_hash", nonExistentBackends[0]+"_0"))
+	reqPerBackend := checkRPCSendOK(ctx, t, client, 1)
+	var got string
+	for got = range reqPerBackend {
+	}
+	if want := backends[0]; got != want {
+		t.Errorf("Got RPC routed to addr %v, want %v", got, want)
+	}
+}
+
+// Tests for a bug seen in the wild in c-core, where ring_hash started with no
+// endpoints and reported TRANSIENT_FAILURE, then got an update with endpoints
+// and reported IDLE, but the picker update was squelched, so it failed to ever
+// get reconnected.
+func (s) TestRingHash_ReattemptWhenGoingFromTransientFailureToIdle(t *testing.T) {
+	const clusterName = "cluster"
+	endpoints := e2e.EndpointResourceWithOptions(e2e.EndpointOptions{
+		ClusterName: clusterName,
+		Localities:  []e2e.LocalityOptions{{}}, // note the empty locality (no endpoint).
+	})
+	cluster := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		ClusterName: clusterName,
+		ServiceName: clusterName,
+		Policy:      e2e.LoadBalancingPolicyRingHash,
+	})
+	route := e2e.DefaultRouteConfig("new_route", virtualHostName, clusterName)
+	listener := e2e.DefaultClientListener(virtualHostName, route.Name)
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	xdsServer, nodeID, xdsResolver := setupManagementServerAndResolver(t)
+	if err := xdsServer.Update(ctx, xdsUpdateOpts(nodeID, endpoints, cluster, route, listener)); err != nil {
+		t.Fatalf("Failed to update xDS resources: %v", err)
+	}
+
+	conn, err := grpc.NewClient("xds:///test.server", grpc.WithResolvers(xdsResolver), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("Failed to create client: %s", err)
+	}
+	defer conn.Close()
+	client := testgrpc.NewTestServiceClient(conn)
+
+	if got, want := conn.GetState(), connectivity.Idle; got != want {
```

Comment:
We could possibly use the internal API to track connectivity states here as well?

Reply:
Let's discuss in https://github.com/grpc/grpc-go/pull/7334/files#r1659181411.